app/models/user_password.rb

```diff
@@ -1,14 +1,67 @@
 # frozen_string_literal: true
 
 class UserPassword < ActiveRecord::Base
-  validates :user_id, presence: true
+  MAX_PASSWORD_LENGTH = 200
+  TARGET_PASSWORD_ALGORITHM =
+    "$pbkdf2-#{Rails.configuration.pbkdf2_algorithm}$i=#{Rails.configuration.pbkdf2_iterations},l=32$"
+  PASSWORD_SALT_LENGTH = 16
+
+  attr_reader :password # required to build an instance of this model with password attribute without backing column
+
+  belongs_to :user, required: true
 
   validates :user_id, uniqueness: true
   validates :password_hash, presence: true, length: { is: 64 }, uniqueness: { scope: :user_id }
   validates :password_salt, presence: true, length: { is: 32 }
   validates :password_algorithm, presence: true, length: { maximum: 64 }
+  validate :password_validator
+
+  def password=(pw)
+    return if pw.blank?
+    @password = pw
+
+    # ensure password is hashed before validations, this should not be in a before_validation hook as that's skipped during .save(validate: false) (e.g. Seedfu)
+    self.password_salt = SecureRandom.hex(PASSWORD_SALT_LENGTH)
+    self.password_algorithm = TARGET_PASSWORD_ALGORITHM
+    self.password_hash = hash_password(password, password_salt, password_algorithm)
+  end
+
+  def password_validation_required?
+    password.present?
+  end
+
+  def confirm_password?(pw)
+    # nothing to confirm if this record has not been persisted yet
+    return false if !persisted?
+
+    if password_hash_changed? &&
+         password_hash_was != hash_password(pw, password_salt_was, password_algorithm_was)
+      return false
+    end
```

Comment:
This is kinda surprising logic. What's it for? It seems like it's performing some validation based on the previous password? 🤔